### PR TITLE
crimson/osd: Snaps IO path

### PR DESCRIPTION
**Pull request by Matan-B**
Creating copy-on-write (COW) clones in the backing store when a write arrives after a snapshot is taken of an object.
Both for Pool or Self-managed snapshotting mechanisms.

Trello: https://trello.com/c/yTDCGNCf
Signed-off-by: Matan Breizman <mbreizma@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/ops_executer.cc [hunk 1]
```diff
@@ -723,6 +724,115 @@ version_t OpsExecuter::get_last_user_version() const
...
+
+    //  clone object, the snap field is set to the seq of the SnapContext
+    //  at it's creation.
+    hobject_t coid = soid;
+    coid.snap = snapc.seq;
+
+    unsigned l;
+    for (l=1;
+         l < snapc.snaps.size() && snapc.snaps[l] > obc->ssc->snapset.seq;
+	 l++) ;
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:746]:
```c++
    const std::vector<snapid_t> snaps = [&] {
      auto last = std::find_if(
        std::begin(snapc.snaps), std::end(snapc.snaps),
        [&](snap_id_t snap_id) { return snap_id > obc->ssc->snapset.seq; });
      return std::vector<snapid_t>{std::begin(snapc.snaps), last};
    }();
```

File: src/crimson/osd/ops_executer.cc [hunk 2]
```diff
@@ -723,6 +724,115 @@ version_t OpsExecuter::get_last_user_version() const
...
+  if (head_existed &&                            // old obs.exists
+      snapc.snaps.size() &&                      // there are snaps
+      snapc.snaps[0] > obc->ssc->snapset.seq) {  // existing onj is old
+
+    //  clone object, the snap field is set to the seq of the SnapContext
+    //  at it's creation.
+    hobject_t coid = soid;
+    coid.snap = snapc.seq;
+
+    unsigned l;
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:743]: please do not use a single `l` for variable name. it is difficult to tell it from `1`, and it deserves a better name.

File: src/crimson/osd/ops_executer.cc [hunk 3]
```diff
@@ -723,6 +724,115 @@ version_t OpsExecuter::get_last_user_version() const
...
+    for (l=1;
+         l < snapc.snaps.size() && snapc.snaps[l] > obc->ssc->snapset.seq;
+	 l++) ;
+
+    std::vector<snapid_t> snaps(l);
+    for (unsigned i=0; i<l; i++) {
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:749]: please follow the code convention. namely, add spaces around operators.

File: src/crimson/osd/ops_executer.cc [hunk 4]
```diff
@@ -723,6 +724,115 @@ version_t OpsExecuter::get_last_user_version() const
...
+      snaps[i] = snapc.snaps[i];
+    }
+
+    // version
+    osd_op_params->at_version = pg->next_version();
+
+    //prepare clone
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:756]: please add a  space after `//`.

File: src/crimson/osd/ops_executer.cc [hunk 5]
```diff
@@ -723,6 +724,115 @@ version_t OpsExecuter::get_last_user_version() const
...
+    // newsnapset is obc's ssc
+    obc->ssc->snapset.clones.push_back(coid.snap);
+    obc->ssc->snapset.clone_size[coid.snap] = obc->obs.oi.size;
+    obc->ssc->snapset.clone_snaps[coid.snap] = snaps;
+
+    // clone_overlap should contain an entry for each clone
+    // (an empty interval_set if there is no overlap)
+    obc->ssc->snapset.clone_overlap[coid.snap];
+    if (obc->obs.oi.size)
+      obc->ssc->snapset.clone_overlap[coid.snap].insert(0, obc->obs.oi.size);
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:796]: I'd suggest always add curly braces around single line statement.

File: src/crimson/osd/pg.cc [hunk 6]
```diff
@@ -847,11 +866,14 @@ std::optional<hobject_t> PG::resolve_oid(
...
 {
+  logger().debug("{} oid.snap={},head snapset.seq={}",
+                 __func__, oid.snap, ss.seq);
   if (oid.snap > ss.seq) {
+    // perfrom a read from head
     return oid.get_head();
   } else {
     // which clone would it be?
-    auto clone = std::upper_bound(
+    auto clone = std::find(
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc:876]: why? if we want to ensure that `clone` is of the same `oid.snap`, shall we use `std::lower_bound()`? do you want to optimize for the case of small `std::size(ss.clones)`?

File: src/crimson/osd/pg_backend.cc [hunk 7]
```diff
@@ -80,47 +80,53 @@ PGBackend::load_metadata(const hobject_t& oid)
...
+        if (auto oiiter = attrs.find(OI_ATTR); oiiter != attrs.end()) {
+          bufferlist bl = std::move(oiiter->second);
+          ret->os = ObjectState(
+            object_info_t(bl, oid),
+            true);
+        } else {
+          logger().error(
+            "load_metadata: object {} present but missing object info",
+            oid);
+          return crimson::ct_error::object_corrupted::make();
```
Comment (tchaikov) [inline src/crimson/osd/pg_backend.cc:93]: in general, i'd avoid blending formatting changes into functional ones.

**Comment by athanatos** [comment]
I'll try to get to this on Monday.

File: src/crimson/osd/ops_executer.cc [hunk 8]
```diff
@@ -723,6 +724,115 @@ version_t OpsExecuter::get_last_user_version() const
...
+    }
+    logger().debug("{} {} done, snapset={}",
+      __func__, soid, obc->ssc->snapset);
+  }
+}
+
+void OpsExecuter::_make_clone()
+{
+  logger().debug("{}", __func__);
+  ++num_write;
```
Comment (rzarzynski) [inline src/crimson/osd/ops_executer.cc:826]: How about using `do_write_op()` for this? The adventage is saying: _"this is a write-like op"_.

File: src/crimson/osd/ops_executer.h [hunk 9]
```diff
@@ -273,6 +281,10 @@ class OpsExecuter : public seastar::enable_lw_shared_from_this<OpsExecuter> {
     const std::vector<OSDOp>& ops);
   void fill_op_params_bump_pg_version();
 
+  const ObjectContextRef get_obc() const {
```
Comment (rzarzynski) [inline src/crimson/osd/ops_executer.h:284]: Hmm, is `const ObjectContextRef` really intentional?

File: src/crimson/osd/pg_backend.cc (hunk 7 above)
Comment (rzarzynski) [reply src/crimson/osd/pg_backend.cc:93]: :+1: A dedicated commit would be perfect.

File: src/crimson/osd/object_context.h [hunk 10]
```diff
@@ -35,14 +39,24 @@ struct obc_to_hoid {
   }
 };
 
+struct SnapSetContext {
+  hobject_t oid;
+  SnapSet snapset;
+  int ref;
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/object_context.h:45]: If the goal here is to refcount SnapSetContext, please use https://www.boost.org/doc/libs/1_46_0/libs/smart_ptr/intrusive_ptr.html and in particular https://www.boost.org/doc/libs/1_62_0/libs/smart_ptr/intrusive_ref_counter.html

File: src/crimson/osd/ops_executer.cc [hunk 11]
```diff
@@ -723,6 +724,115 @@ version_t OpsExecuter::get_last_user_version() const
   return pg->get_last_user_version();
 }
 
+void OpsExecuter::make_writeable(std::vector<pg_log_entry_t>& log_entries) {
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/ops_executer.cc:727]: In general, opening brace goes on its own line in functions/methods defined outside of class/struct definitions and on the same line for methods defined within the class definition.

**Comment by athanatos** [comment]
I'll take another look once the above have been addressed.

File: src/crimson/osd/pg.cc (hunk 6 above)
Comment (Matan-B) [reply src/crimson/osd/pg.cc]: Changed to `lower_bound()` for now. This optimization seems to be negligible.

File: src/crimson/osd/pg_backend.cc (hunk 7 above)
Comment (Matan-B) [reply src/crimson/osd/pg_backend.cc:93]: 5184d2903beddee04bb16ff6d9995a5b90693fb2

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
> I'll take another look once the above have been addressed.

@athanatos They have been addressed.

File: src/crimson/osd/ops_executer.cc [hunk 12]
```diff
@@ -723,6 +724,114 @@ version_t OpsExecuter::get_last_user_version() const
...
+      return std::vector<snapid_t>{std::begin(snapc.snaps), last};
+    }();
+
+    // version
+    osd_op_params->at_version = pg->next_version();
+
+    auto snap_oi = prepare_clone(coid);
+
+    // make clone
+    do_write_op([this] (auto& backend, auto& os, auto& txn) {
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.cc:759]: ](

File: src/crimson/osd/ops_executer.cc [hunk 13]
```diff
@@ -723,6 +724,114 @@ version_t OpsExecuter::get_last_user_version() const
...
+      return backend.clone(os, clone_obc->obs, txn);
+    }, true);
+    // TODO: maybe cache:
+    // encode(snap_oi, bv, get_osdmap()->get_features(CEPH_ENTITY_TYPE_OSD, nullptr));
+    // txn.setattr(coll->get_cid(), ghobject_t{obc->obs.oi.soid}, OI_ATTR, bv);
+    // txn.rmattr(coll->get_cid(), ghobject_t{obc->obs.oi.soid}, SS_ATTR);
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.cc:765]: This probably needs to be uncommented?

File: src/crimson/osd/ops_executer.cc [hunk 14]
```diff
@@ -723,6 +724,114 @@ version_t OpsExecuter::get_last_user_version() const
...
+    // newsnapset is obc's ssc
+    obc->ssc->snapset.clones.push_back(coid.snap);
+    obc->ssc->snapset.clone_size[coid.snap] = obc->obs.oi.size;
+    obc->ssc->snapset.clone_snaps[coid.snap] = cloned_snaps;
+
+    // clone_overlap should contain an entry for each clone
+    // (an empty interval_set if there is no overlap)
+    if (obc->obs.oi.size) {
+      obc->ssc->snapset.clone_overlap[coid.snap].insert(0, obc->obs.oi.size);
+    }
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.cc:781]:
```
    auto &overlap = obc->ssc->snapset.clone_overlap[coid.snap];
    if (obc->obs.oi.size) {
      overlap.insert(0, obc->obs.oi.size);
    }
```
since you need to create an empty map at clone_overlap[coid.snap] unconditionally?

File: src/crimson/osd/ops_executer.cc [hunk 15]
```diff
@@ -723,6 +724,114 @@ version_t OpsExecuter::get_last_user_version() const
...
+
+    log_entries.emplace_back(pg_log_entry_t::CLONE,
+                             coid, osd_op_params->at_version,
+                             obc->obs.oi.version, obc->obs.oi.user_version,
+                             osd_reqid_t(),
+			     obc->obs.oi.mtime, 0);
+    encode(cloned_snaps, log_entries.back().snaps);
+    osd_op_params->at_version.version++;
+
+    // TODO: update most recent clone_overlap and usage stats
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.cc:796]: Is this for a future PR?

File: src/crimson/osd/ops_executer.cc [hunk 16]
```diff
@@ -723,6 +724,114 @@ version_t OpsExecuter::get_last_user_version() const
...
+      auto [c_obc, existed] =
+        pg->get_shard_services().obc_registry.get_cached_obc(
+          std::move(coid));
+      assert(!existed);
+      c_obc->obs.oi = static_snap_oi;
+      c_obc->obs.exists = true;
+      c_obc->ssc = obc->ssc;
+      c_obc->head = obc->head;
+      logger().debug("clone_obc:{}", c_obc->obs.oi);
+      clone_obc=std::move(c_obc);
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.cc:823]: spaces around the =

File: src/crimson/osd/ops_executer.cc [hunk 17]
```diff
@@ -723,6 +724,114 @@ version_t OpsExecuter::get_last_user_version() const
...
+      // lookup_or_create
+      auto [c_obc, existed] =
+        pg->get_shard_services().obc_registry.get_cached_obc(
+          std::move(coid));
+      assert(!existed);
+      c_obc->obs.oi = static_snap_oi;
+      c_obc->obs.exists = true;
+      c_obc->ssc = obc->ssc;
+      c_obc->head = obc->head;
+      logger().debug("clone_obc:{}", c_obc->obs.oi);
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.cc:822]: clone_obc: {}

space after :

File: src/crimson/osd/ops_executer.cc (hunk 14 above)
Comment (Matan-B) [reply src/crimson/osd/ops_executer.cc:781]: The map will be created once the operator[] will be used. Why should we create it unconditionally?

File: src/crimson/osd/ops_executer.cc (hunk 15 above)
Comment (Matan-B) [reply src/crimson/osd/ops_executer.cc:796]: Yes, once further snaps parts will be implemented.

File: src/crimson/osd/ops_executer.cc (hunk 13 above)
Comment (Matan-B) [reply src/crimson/osd/ops_executer.cc]: Updated.

Comment (Matan-B) [reply src/crimson/osd/ops_executer.cc]: > Getting really close! What testing have you done so far?

Mostly local testing using the rados tool. 
> If you've got some teuthology tests you used, I think the next step would be to split those out into a crimson smoke suite and start gating all crimson PRs on it.

I agree, I will start working on a minimal suite to run against.

**Comment by Matan-B** [comment]
> Getting really close! What testing have you done so far?

Mostly local testing using the rados tool. 
> If you've got some teuthology tests you used, I think the next step would be to split those out into a crimson smoke suite and start gating all crimson PRs on it.

I agree, I will start working on a minimal suite to run against.

File: src/crimson/osd/ops_executer.cc (hunk 14 above)
Comment (Matan-B) [reply src/crimson/osd/ops_executer.cc:777]: [aea0cb9be2b958dd9750add70c3b73ebbee64bcd](https://github.com/ceph/ceph/pull/46941/commits/aea0cb9be2b958dd9750add70c3b73ebbee64bcd#diff-04d5a3324b706d64a3230221d2eb08f6aa66829f93ca151ebb71896fb743ada7L8628) edited also here, will update accordingly.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/pg.cc [hunk 18]
```diff
@@ -786,11 +801,13 @@ PG::do_osd_ops(
   do_osd_ops_success_func_t success_func,
   do_osd_ops_failure_func_t failure_func)
 {
-  return seastar::do_with(std::move(msg_params), [=, &ops, &op_info]
+  //internal client request
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:804]:
```
  /* This overload is generally used for internal client requests, use an empty
  * SnapContext */
```

File: src/crimson/osd/pg.cc [hunk 19]
```diff
@@ -786,11 +801,13 @@ PG::do_osd_ops(
...
+  SnapContext snapc;  //empty snapc
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:805]: Don't capture stack variables by reference unnecessarily.  Remove this declaration and modify below to:

```
        Ref<PG>{this}, std::move(obc), op_info, msg_params, SnapContext{}),
```

Also, reformat to:

```
  return seastar::do_with(
    std::move(msg_params),
    [=, &ops, &op_info, &snapc](auto &msg_params) {
      return do_osd_ops_execute<void>(
        seastar::make_lw_shared<OpsExecuter>(
          Ref<PG>{this}, std::move(obc), op_info, msg_params, SnapContext{}),
        ops,
        std::move(success_func),
        std::move(failure_func));
    });
```

File: src/crimson/osd/pg.cc [hunk 20]
```diff
@@ -930,12 +949,11 @@ PG::with_clone_obc(hobject_t oid, with_obc_func_t&& func)
     [oid, func=std::move(func), this](auto head) -> load_obc_iertr::future<> {
     auto coid = resolve_oid(head->get_ro_ss(), oid);
     if (!coid) {
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:951]: This change to permit returning an object_corrupted should be in its own commit since it seems to be independent of the rest of the commit.

File: src/crimson/osd/pg.cc [hunk 21]
```diff
@@ -847,22 +864,24 @@ std::optional<hobject_t> PG::resolve_oid(
   const SnapSet &ss,
   const hobject_t &oid)
 {
+  logger().debug("{} oid.snap={},head snapset.seq={}",
+                 __func__, oid.snap, ss.seq);
   if (oid.snap > ss.seq) {
+    // perfrom a read from head
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:870]: Either remove this comment or make it more explanatory:
```
    // Because oid.snap > ss.seq, we are trying to read from a snapshot
    // taken after the most recent write to this object.  Read from head.
```

File: src/crimson/osd/pg.cc [hunk 22]
```diff
@@ -847,22 +864,24 @@ std::optional<hobject_t> PG::resolve_oid(
...
+    auto clone = std::lower_bound(
       begin(ss.clones), end(ss.clones),
       oid.snap);
     if (clone == end(ss.clones)) {
       // Doesn't exist, > last clone, < ss.seq
       return std::nullopt;
     }
     auto citer = ss.clone_snaps.find(*clone);
-    // TODO: how do we want to handle this kind of logic error?
-    ceph_assert(citer != ss.clone_snaps.end());
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:863]: If citer == ss.clone_snaps.end(), we have an entry in ss.clones without a corresponding ss.clone_snaps entry.  That is meant to be impossible.  For now, an assert is fine.  In the future, we'd probably want to propagate some kind of logic error so scrub can try to fix it.

File: src/crimson/osd/ops_executer.h [hunk 23]
```diff
@@ -166,16 +166,22 @@ class OpsExecuter : public seastar::enable_lw_shared_from_this<OpsExecuter> {
...
   std::optional<osd_op_params_t> osd_op_params;
   bool user_modify = false;
+  bool head_existed = false;
   ceph::os::Transaction txn;
 
   size_t num_read = 0;    ///< count read ops
   size_t num_write = 0;   ///< count update ops
 
+  SnapContext snapc; // writer snap context
+  // obc will hold both (old/resulting)obs
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.h:182]: I'm not sure this context is useful.  obc unconditionally references the snapcontext even for a read, right?  This also doesn't seem to be a good place to explain that the obc contains the obs.

File: src/crimson/osd/ops_executer.cc [hunk 24]
```diff
@@ -723,6 +724,117 @@ version_t OpsExecuter::get_last_user_version() const
...
+      assert(!existed);
+      c_obc->obs.oi = static_snap_oi;
+      c_obc->obs.exists = true;
+      c_obc->ssc = obc->ssc;
+      c_obc->head = obc->head;
+      logger().debug("clone_obc: {}", c_obc->obs.oi);
+      clone_obc = std::move(c_obc);
+      snap_oi = &clone_obc->obs.oi;
+    } else {
+      snap_oi = &static_snap_oi;
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.cc:823]: You can't return a pointer to an automatic variable, it will have been destructed before the caller sees it.

File: src/crimson/osd/pg_backend.cc [hunk 25]
```diff
@@ -80,47 +80,53 @@ PGBackend::load_metadata(const hobject_t& oid)
...
+            logger().debug(
+              "load_metadata: object {} and snapset {} present",
+              oid, ssiter->second);
+            bufferlist bl = std::move(ssiter->second);
+            ret->ssc = new crimson::osd::SnapSetContext(oid.get_snapdir());
+            ret->ssc->snapset = SnapSet(bl);
+            if (bl.length()) {
+              ret->ssc->exists = true;
+            } else {
+              ret->ssc->exists = false;
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:107]: What does SnapSetContext::exists mean?  IIRC, it's there because a cached SnapSetContext may persist after the logical object to which it belongs is deleted.  If it happens to still be there when the object is read, get_snapset_context will notice and return NULL rather than the context.  For a write, the same context will be returned with exists=false, just as if it were newly created.

PrimaryLogPG::get_snapset_context afaict is using (abusing?) bv.empty() to indicate that the object doesn't exists and can_create.  The bv.length() check should be inaccessible unless either the object exists and bv.length() > 0 or it doesn't exist and can_create in which case bv.length() == 0. (Let me know if I've misread this).

This check is specifically in the path where the object really exists -- the enoent case below handles the object missing case.  Thus, if bl.length() == 0 (or can't be decoded into a SnapSet for any other reason), you should be returning object_corrupted, not an empty context with exists = false.  Also, SnapSet(bl) should be in a try{} block so you can notice a decode error and return an error accordingly.

File: src/crimson/osd/pg_backend.cc [hunk 26]
```diff
@@ -80,47 +80,53 @@ PGBackend::load_metadata(const hobject_t& oid)
...
-	  });
+        logger().debug(
+          "load_metadata: object {} doesn't exist, returning empty metadata",
+          oid);
+        return load_metadata_ertr::make_ready_future<loaded_object_md_t::ref>(
+          new loaded_object_md_t{
+            ObjectState(
+              object_info_t(oid),
+              false),
+            oid.is_head() ? (new crimson::osd::SnapSetContext(oid)) : nullptr
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:128]: This returned SnapSetContext should have exists=false, I think.

**Review by athanatos — CHANGES_REQUESTED** [review]
Did another pass, re-request review when ready.  Getting close!

File: src/crimson/osd/ops_executer.cc [hunk 27]
```diff
@@ -723,6 +724,117 @@ version_t OpsExecuter::get_last_user_version() const
...
+  if (head_existed &&                            // old obs.exists
+      snapc.snaps.size() &&                      // there are snaps
+      snapc.snaps[0] > obc->ssc->snapset.seq) {  // existing obj is old
+
+    //  clone object, the snap field is set to the seq of the SnapContext
+    //  at its creation.
+    hobject_t coid = soid;
+    coid.snap = snapc.seq;
+
+  // existing snaps are stored desendingly in snapc,
```
Comment (ronen-fr) [inline src/crimson/osd/ops_executer.cc:744]: sp?

File: src/crimson/osd/pg_backend.cc (hunk 25 above)
Comment (Matan-B) [reply src/crimson/osd/pg_backend.cc]: Should I push 7a03a88eca32fac4a29d789ba651d989a63a81a7 as a separate PR as it involves classic osd?

File: src/crimson/osd/object_context.h [hunk 28]
```diff
@@ -35,14 +37,26 @@ struct obc_to_hoid {
   }
 };
 
+struct SnapSetContext :
+  public boost::intrusive_ref_counter<SnapSetContext,
+                                     boost::thread_unsafe_counter>
+{
+  hobject_t oid;
+  SnapSet snapset;
+  bool exists;
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/object_context.h:46]: Something like

```
/**
 * exists
 *
 * Because ObjectContext's are cached, we need to be able to express the case
 * where the object to which a cached ObjectContext refers does not exist.
 * ObjectContext's for yet-to-be-created objects are initialized with exists=false.
 * The ObjectContext for a deleted object will have exists set to false until it falls
 * out of cache (or another write recreates the object).
 */
```

**Comment by athanatos** [comment]
The information from your PR description needs to be in a commit message -- presumably 'crimson/osd: Snapshots: copy-on-write clones'

File: src/crimson/osd/ops_executer.cc (hunk 24 above)
Comment (athanatos) [reply src/crimson/osd/ops_executer.cc:824]: You still seem to be returning a pointer to a stack variable.  Am I missing something?

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/ops_executer.cc [hunk 29]
```diff
@@ -724,6 +725,117 @@ version_t OpsExecuter::get_last_user_version() const
...
+      snapc.snaps[0] > obc->ssc->snapset.seq) {  // existing obj is old
+
+    //  clone object, the snap field is set to the seq of the SnapContext
+    //  at its creation.
+    hobject_t coid = soid;
+    coid.snap = snapc.seq;
+
+  // existing snaps are stored in descending order in snapc,
+  // cloned_snaps vector will hold all the snaps stored until snapset.seq
+   std::vector<snapid_t> cloned_snaps = [&] {
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/osd/ops_executer.cc:747]: also, wrong indent.

File: src/crimson/osd/ops_executer.cc [hunk 30]
```diff
@@ -724,6 +725,117 @@ version_t OpsExecuter::get_last_user_version() const
...
+       obc->ssc->snapset.snaps.clear();
+    }
+    logger().debug("{} {} done, snapset={}",
+      __func__, soid, obc->ssc->snapset);
+  }
+}
+
+object_info_t* OpsExecuter::prepare_clone(const hobject_t& coid)
+{
+    object_info_t static_snap_oi(coid);
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.cc:808]: wrong indent.

File: src/crimson/osd/ops_executer.h [hunk 31]
```diff
@@ -305,6 +315,16 @@ class OpsExecuter : public seastar::enable_lw_shared_from_this<OpsExecuter> {
...
 
   version_t get_last_user_version() const;
+
+  const SnapContext& get_snapc() const {
+    return snapc;
+  }
+
+  void make_writeable(std::vector<pg_log_entry_t>& log_entries);
+
+  object_info_t* prepare_clone(const hobject_t& coid);
```
[the remark this answers is not in the document]
Comment (Matan-B) [reply src/crimson/osd/ops_executer.h]: Added.
```
  // For primary osd, the returned value will point to the newly
  // created clone object info. Otherwise, to an on-stack static snap.
  object_info_t* prepare_clone(
    const hobject_t& coid,
    object_info_t& static_snap_oi);
```

**Comment by Matan-B** [comment]
* > The information from your PR description needs to be in a commit message -- presumably 'crimson/osd: Snapshots: copy-on-write clones'

  @athanatos, Added.

* All comments have been addressed.
f3d1d5397f218dadf6acfd3bcd0c6d041f34e491 is still WIP, snapset_contexts map will be implemented as we need a way to obtain relevant ssc when the head object is not available (removed). We can use `head->get_ro_ss()` only when the head exists.

File: src/crimson/osd/object_context.h [hunk 32]
```diff
@@ -26,6 +26,8 @@ namespace crimson::common {
 namespace crimson::osd {
 
 class Watch;
+struct SnapSetContext;
+using SnapContextRef = boost::intrusive_ptr<SnapSetContext>;
```
Comment (athanatos) [inline src/crimson/osd/object_context.h:30]: SnapSetContextRef

File: src/crimson/osd/object_context.h [hunk 33]
```diff
@@ -35,14 +37,34 @@ struct obc_to_hoid {
   }
 };
 
+struct SnapSetContext :
+  public boost::intrusive_ref_counter<SnapSetContext,
+                                     boost::thread_unsafe_counter>
+{
+  hobject_t oid;
+  SnapSet snapset;
+  bool exists;
```
Comment (athanatos) [inline src/crimson/osd/object_context.h:46]: bool exists = false;

File: src/crimson/osd/object_context.h [hunk 34]
```diff
@@ -69,18 +91,17 @@ class ObjectContext : public ceph::common::intrusive_lru_base<
...
+      ceph_assert(ssc);
+      return ssc->snapset;
     } else {
-      ceph_assert(head);
       return head->get_ro_ss();
     }
   }
 
-  void set_head_state(ObjectState &&_obs, SnapSet &&_ss) {
+  void set_head_state(ObjectState &&_obs, SnapContextRef&& _ssc) {
```
Comment (athanatos) [inline src/crimson/osd/object_context.h:101]: SnapContextRef &&_ssc

File: src/crimson/osd/pg.cc [hunk 35]
```diff
@@ -1031,26 +1053,34 @@ PG::with_clone_obc(hobject_t oid, with_obc_func_t&& func)
   assert(!oid.is_head());
   return with_head_obc<RWState::RWREAD>(oid.get_head(),
     [oid, func=std::move(func), this](auto head) -> load_obc_iertr::future<> {
-    auto coid = resolve_oid(head->get_ro_ss(), oid);
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:1034]: We don't really want to copy the snapset here, how about:

```
if (!head->obs.exists) {
  // WIP: Introduce snapset_contexts map in shard_services.
  logger().error("with_clone_obc: head ssc not found");
  return load_obc_iertr::future<>{crimson::ct_error::object_corrupted::make()};
}
auto &snapset = head->get_ro_ss()
...
```

File: src/crimson/osd/pg_backend.cc [hunk 36]
```diff
@@ -80,47 +80,68 @@ PGBackend::load_metadata(const hobject_t& oid)
...
-	}
+        loaded_object_md_t::ref ret(new loaded_object_md_t());
+        if (auto oiiter = attrs.find(OI_ATTR); oiiter != attrs.end()) {
+          bufferlist bl = std::move(oiiter->second);
+          try {
+            ret->os = ObjectState(
+              object_info_t(bl, oid),
+              true);
+          } catch (const buffer::error&) {
+                logger().warn("unable to decode ObjectState");
```
Comment (athanatos) [inline src/crimson/osd/pg_backend.cc:91]: spacing looks inconsistent here

**Review by athanatos — APPROVED** [review]
LGTM other than the above nits.  I'm out next week, I'll be back on the 16th.

**Comment by Matan-B** [comment]
737f39556a9fb6fc779c7bc30bb137a188bc977e depends on #47554. (Before we can run it against Teuthology)
Local results:
```
test_rados.TestIoctx.test_create_snap ... ok                                                                                                                                                  
test_rados.TestIoctx.test_list_snaps ... ok
test_rados.TestIoctx.test_list_snaps_empty ... ok
test_rados.TestIoctx.test_lookup_snap ... ok
test_rados.TestIoctx.test_remove_snap ... ok
test_rados.TestIoctx.test_snap_read ... ok
test_rados.TestIoctx.test_snap_timestamp ... ok
test_rados.TestIoctx.test_remove_snap ... ok
test_rados.TestIoctx.test_snap_read ... ok
test_rados.TestIoctx.test_snap_timestamp ... ok
```

**Comment by athanatos** [comment]
retest this please

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test api

**Comment by Matan-B** [comment]
jenkins test api

**Comment by rzarzynski** [comment]
This is the very last test that prevents `TestClsRbd` to be entirely green on my machine:

```
[----------] Global test environment tear-down
[==========] 69 tests from 1 test suite ran. (6711 ms total)
[  PASSED  ] 68 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TestClsRbd.assert_snapc_seq
```

**Comment by Matan-B** [comment]
> `assert_snapc_seq` is the very last test that prevents `TestClsRbd` to be entirely green on my machine:
> 
> ```
> [----------] Global test environment tear-down
> [==========] 69 tests from 1 test suite ran. (6711 ms total)
> [  PASSED  ] 68 tests.
> [  FAILED  ] 1 test, listed below:
> [  FAILED  ] TestClsRbd.assert_snapc_seq
> ```

Addressed in this PR.

**Comment by Matan-B** [comment]
Merging as discussed in Q & A.
